drivers/ws281x/timer_gpio_ll.c

```diff
@@ -0,0 +1,162 @@
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+/* + NS_PER_SEC - 1: Rounding up, as this may be longer */
+const int ticks_one = ((uint64_t)WS281X_T_DATA_ONE_NS * (uint64_t)WS281X_TIMER_FREQ + NS_PER_SEC - 1)/ NS_PER_SEC;
+/* Rounding down, zeros are better shorter */
+const int ticks_zero = (uint64_t)WS281X_T_DATA_ZERO_NS * (uint64_t)WS281X_TIMER_FREQ / NS_PER_SEC;
+/* No particular known requirements, but we're taking longer than that anyway
+ * because we don't clock the times between bits. */
+const int ticks_data = (uint64_t)WS281X_T_DATA_NS * (uint64_t)WS281X_TIMER_FREQ / NS_PER_SEC;
+
+void ws281x_write_buffer(ws281x_t *dev, const void *buf, size_t size)
+{
+    assert(dev);
+
+    const uint8_t *pos = buf;
+    const uint8_t *end = pos + size;
+
+    /* The logical sequence for a cycle would be to set high, sleep, set low,
+     * sleep. But the critical time is the high time, so we do something
+     * different: We go low, set up the timer, and then go high and back low
+     * both as controlled by the timer. That means that our logical cycles are
+     * the previous bit's low time followed by the current bit's high time -- so we store the previous bit here.
+     *
+     * It is initialized to 1 because the 1 has the longer high time, so the
+     * lower low time -- we just spend 325 instead of 600ns before actually
+     * sending the message. (This wait-before-we-start time is also part of how
+     * this implementation can get away without compensation for the time it
+     * takes to set up timers: it sets them up and then waits for some time
+     * that is known to be long enough, because that time also works
+     * mid-transmission). */
+    bool last_bit = 1;
+
+    gpio_port_t port = gpio_get_port(dev->params.pin);
+    uword_t mask = 1 << gpio_get_pin_num(dev->params.pin);
+
+    /* If a too-slow channel is used, then the critical section could be done
+     * away with: If an interrupt fires (or a higher priority thread runs;
+     * really: runs long enough to matter, but that'd need very fast switching
+     * and hundreds of MHz to be the case), the function will return with an
+     * error (or ignore the error), and the worst thing that'd happen is that
+     * if it hits the last bit of a LED, that might be flipped to a value 1.
+     * */
+    unsigned int irq_state = irq_disable();
+
+    while (pos < end) {
+        uint8_t data = *pos;
+        for (uint8_t cnt = 8; cnt > 0; cnt--) {
+            bool bit = data >> 7;
+            data <<= 1;
+            int time_to_set = last_bit ? (ticks_data - ticks_one) : (ticks_data - ticks_zero);
+            int time_to_clear = time_to_set + (bit ? ticks_one : ticks_zero);
+
+            /* I'd prefer to just zero the timer, but we don't have API for that */
+            int now = timer_read(WS281X_TIMER_DEV);
+            time_to_set = (time_to_set + now) & WS281X_TIMER_MAX_VALUE;
+            time_to_clear = (time_to_clear + now) & WS281X_TIMER_MAX_VALUE;
```

Comment:
Sadly the timer period is variable rather than fixed (and in fact is what is used to encode the bits)

Reply:
I had hoped the periods would have some common divider

Reply:
325 ns should work fine:

https://github.com/RIOT-OS/RIOT/blob/master/drivers/ws281x/include/ws281x_constants.h#L39-L45

With 4 times 325 ns being close to the total bit time, this would likely be in spec (despite the ws281x actually being pretty forgiving there). I guess this could work quite well, provided timer IRQs are not disabled for more than 100 ns.

Reply:
arg, that will be too short for most timers

Reply:
Yeah, but that issue also applies equally to not using periodic timers, as the 325 ns high time also needs to be implemented there as well.

Reply:
It's one thing to be short for timers, but it's another that ISRs are too slow. On 64MHz devices there are like 20 cycles for this to complete. Something in the order of 12 cycles goes just for the ISR to enter, then even if it's just 1 cycle because we hard-code which pin to toggle, but we still need to get out of the ISR and that needs to be done before the next interrupt fires on the same interrupt level.